tests/cpp/python/python_client_tests.py

```diff
@@ -0,0 +1,212 @@
+                "Server: EXASolution R Package\r\n" \
+                "Content-type: application/octet-stream\r\n" \
+                "Content-disposition: attachment; filename=data.csv\r\n" \
+                "Connection: close\r\n\r\n"
+    header_from_client = clientsocket.recv(len(ok_answer))
+    assert header_from_client == ok_answer.encode("UTF-8")
+    data = clientsocket.recv(3)
+    assert data == "\"a\"".encode("UTF-8")
+    data = clientsocket.recv(1)
+    assert data == "\n".encode("UTF-8")
+    data = clientsocket.recv(3)
+    assert data == "\"b\"".encode("UTF-8")
+    data = clientsocket.recv(1)
+    assert data == "\n".encode("UTF-8")
+
+    socket_wrapper.unwrap(clientsocket, protocol)
+
+    (newclientsocket, address) = serversocket.accept()
+    recvMetaInfoRequest = newclientsocket.recv(100)
+    b = bytearray(b'\x00\x00\x00\x00\4\0\0\0Test\0\0\0\0\0\0\0\0\0\0\0\0')
+    newclientsocket.send(b)
+
+    newclientsocket = socket_wrapper.wrap(newclientsocket, protocol)
+    b = bytearray(b"\r\n")  # empty header
+    newclientsocket.send(b)
+    data = "Name\na\nb"
+    b = bytearray(f'{hex(len(data))}\n', 'UTF-8')
+    newclientsocket.send(b)
+    d = bytearray(data, 'UTF-8')
+    d.append(0)
+    d.append(0)
+    newclientsocket.send(d)
+
+    # Send zer termination
+    b = bytearray(f'{0}\n', 'UTF-8')
+    newclientsocket.send(b)
+
+    recvmsg = newclientsocket.recv(100)
+    assert recvmsg == b'HTTP/1.1 200 OK\r\nServer: EXASolution R Package\r\nConnection: close\r\n\r\n'
+
+    socket_wrapper.unwrap(newclientsocket, protocol)
+    p_unit_test.wait()
+    try:
+        clientsocket.shutdown(socket.SHUT_RDWR)
+        newclientsocket.close()
+    except socket.error:
+        pass # Under MacOS shutdown can throw an exception if the client has closed the socket already
+    assert p_unit_test.returncode == 0
+
+
+def httpTest(protocol: string):
+    serversocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    serversocket.bind(("localhost", 5000))
+    # become a server socket
+    serversocket.listen(5)
+
+    p_unit_test = subprocess.Popen(["./r_exasol", "Protocol" + protocol])
+
+    (clientsocket, address) = serversocket.accept()
+    clientsocket = socket_wrapper.wrap(clientsocket, protocol)
+
+    class TestHttpConnection(http.client.HTTPConnection):
+
+        def _mock_connection(self, _address, _timeout, _source_address):
+            return self._mocked_socket
+
+        def __init__(self, s, host: str):
+            super().__init__(host)
+            self._mocked_socket = s
+            self._create_connection = self._mock_connection
+
+    testClient = TestHttpConnection(clientsocket, "localhost")
+    testClient.connect()
+    testClient.putrequest('PUT', '/put')
+    testClient.putheader('Transfer-Encoding', 'chunked')
+    testClient.putheader('Connection', 'Keep-Alive')
+    testClient.putheader('Cache-Control', 'no-cache')
+    testClient.endheaders()
+    data = 'CHUNK DATA;' * 20
```

Comment:
Really, not much of an abstraction with the library :wink:

Reply:
Yeap :)